src/Covenant/ASG.hs

```diff
@@ -234,14 +278,135 @@ lam Scope f = do
 -- @since 1.0.0
 letBind ::
   forall (args :: Natural) (binds :: Natural) (m :: Type -> Type).
-  (MonadHashCons Id ASGNode m) =>
+  (MonadHashCons Id ASGNode m, MonadError ASGCompileError m) =>
+  -- | The type of the let binding
+  TyASGNode ->
   Scope args binds ->
   Ref ->
   (Scope args (binds + 1) -> m Ref) ->
   m Id
-letBind Scope r f = do
-  res <- f Scope
-  refTo . LetInternal r $ res
+letBind letTy scope r f = do
+  let scope' = pushLetToScope letTy scope
+  res <- f scope'
+  refTo (LetInternal letTy r res)
+
+-- Helpers
+
+pushArgToScope ::
+  forall (n :: Natural) (m :: Natural).
+  TyASGNode ->
+  Scope n m ->
+  Scope (n + 1) m
+pushArgToScope ty (Scope args lets) =
+  Scope (Vector.cons ty args) lets
+
+pushLetToScope ::
+  forall (n :: Natural) (m :: Natural).
+  TyASGNode ->
+  Scope n m ->
+  Scope n (m + 1)
+pushLetToScope ty (Scope args lets) =
+  Scope args (Vector.cons ty lets)
+
+-- Typing Helpers
+
+liftTypeError ::
+  (MonadHashCons Id ASGNode m, MonadError ASGCompileError m) =>
+  Either TypeError a -> m a
+liftTypeError e =
+  let e' = case e of
+        Left x -> Left $ ATypeError x
+        Right x -> Right x
+   in liftEither e'
```

Comment:
This is over-constrained: `MonadHashCons` isn't doing anything there as far as I can tell. Surely this is just `liftEither`?

Reply:
Simplified this, removed the redundant constraint